### ISSUE TITLE
Test for unescaped dots in exclusion yields false positives

### ISSUE THREAD

**Issue by fuglede** [EFForg/https-everywhere#7320]
We have [a test](https://github.com/EFForg/https-everywhere/blob/master/utils/trivial-validate.py#L86) whose goal is to catch unescaped dots in exclusions, but as far as I can tell, it does not quite work since `unescaped_dot` is set up to ignore everything that comes after the first slash (which makes sense for `test_unescaped_dots`), so, effectively what happens is that for the exclusion pattern `^http://example.com`, we will only check the `^http:` part of it.

    In [5]: unescaped_dot('^http://example.com')
    Out[5]: False

**Comment by jeremyn** [comment]
I'm confused by the problem. I see nothing wrong with the exclusions you've pointed out in https://github.com/EFForg/https-everywhere/pull/7314#pullrequestreview-3757967 , and also I regularly see tests fail because of unescaped dots.

**Comment by fuglede** [comment]
The third one is unescaped. The tests currently correctly detect unescaped dots for pre-slash `from` attributes in `rule` tags, and slash-less exclusion patterns, but will fail for exclusion patterns such as the one above, as [`unescaped dot`](https://github.com/EFForg/https-everywhere/blob/master/utils/trivial-validate.py#L56) only checks the string `^http:` for unescaped dots in that case.

**Comment by jeremyn** [comment]
The third dot is supposed to be unescaped because it represents a wildcard, such as for `forum`, not a literal period.

I'm really confused. Are you saying Travis did fail for that exclusion? That it didn't fail, but should? I don't know what "false positive" means in this context.

**Comment by fuglede** [comment]
No, the Travis test passed in #7314 but the test is there to catch unescaped dots in exclusion patterns (which should be replaced by more specific character groups).

**Comment by fuglede** [comment]
To make the unintendedness of the current behaviour more obvious, let's make a single addition and see what comes out of if;

```python
def unescaped_dot(s):
    escaped = False
    bracketed = False
    for c in s:
        print(c)
        if c == "\\":
           escaped = not escaped
        elif not escaped and c == "[":
           bracketed = True
        elif not escaped and c == "]":
           bracketed = False
        elif not escaped and not bracketed and c == ".":
           return True
        elif not bracketed and c == "/":
           break
        else:
           escaped = False
    return False
```

Now,
```
In [3]: unescaped_dot('^http://example.com')
^
h
t
t
p
:
/
Out[3]: False
```

As an example of the intended behaviour, let us try with [this exclusion](https://github.com/EFForg/https-everywhere/blob/master/src/chrome/content/rules/CJS-CDKeys.com.xml#L12) which contains no non-bracketed slashes;

```
In [4]: unescaped_dot('^(?![\w:/.-]+\.(css|gif|ico|jpg|js|png)$)')
^
(
?
!
[
\
w
:
/
.
-
]
+
\
.
(
c
s
s
|
g
i
f
|
i
c
o
|
j
p
g
|
j
s
|
p
n
g
)
$
)
Out[4]: False
```

Had we replaced the character group with an unescaped dot, the test would be complaining about the ruleset (as you can see in action by making the change locally and running the script)

```
In [5]: unescaped_dot('^.+\.(css|gif|ico|jpg|js|png)$)')
^
.
Out[5]: True
```

**Comment by jeremyn** [comment]
Got it, thanks. I suppose the times I've seen the unescaped dot error have been with `test_unescaped_dots`, which [strips the slashes](https://github.com/EFForg/https-everywhere/blob/72b1167f773cb4339bebc41fc5abd2429ca69b4a/utils/trivial-validate.py#L81).

**Comment by pipboy96** [comment]
@fuglede @jeremyn Is this issue still relevant?

**Comment by fuglede** [comment]
I didn't check if its usage has changed, but the function itself behaves exactly the same (except it's in [`test/validations/special/run.py`](https://github.com/EFForg/https-everywhere/blob/57a41c2edd22e4ee60c61691dba3c210085ee1e8/test/validations/special/run.py#L54) now).
